Drivers/Src/MotorController.c

```diff
@@ -27,39 +86,48 @@ void MotorController_Init(){
  * @return  None
  */ 
 void MotorController_Drive(float newVelocity, float motorCurrent){
+    CPU_TS ts;
+	OS_ERR err;
 
-    uint32_t nv = *((uint32_t *)((void *) &newVelocity));
-    uint32_t mc = *((uint32_t *)((void *) &motorCurrent));
-
-    uint8_t data[8] = {0};
-    int index = 0;
-    while(index < MAX_CAN_LEN/2){
-        data[index] = (mc >> (8 * (MAX_CAN_LEN/2-index-1))) & 0xFF; //split inputs into bytes
-        index++;
-        
-    }
-    int i = 0;
-    while(index < MAX_CAN_LEN){
-        data[index] = (nv >> (8 * (MAX_CAN_LEN/2-i-1))) & 0xFF;
-        index++;
-        i++;
-    }
+    uint8_t data[8];
+    memcpy(data, &motorCurrent, sizeof(motorCurrent));
+    memcpy(data + sizeof(motorCurrent), &newVelocity, sizeof(newVelocity));
     
-    BSP_CAN_Write(CAN_3, MOTOR_DRIVE, data, MAX_CAN_LEN);
+    OSSemPend(&MotorController_MailSem4,
+			  0,
+			  OS_OPT_PEND_BLOCKING,
+			  &ts,
+			  &err);
+	assertOSError(0, err);
+    ErrorStatus result = BSP_CAN_Write(CAN_3, MOTOR_DRIVE, data, MAX_CAN_LEN);
+    if (result == ERROR) {
+		MotorController_Release();
+	}
+
 }
 
 /**
- * @brief   Reads most recent command from CAN2 bus
+ * @brief   Reads most recent command from CAN3 bus
  * @param   message the buffer in which the info for the CAN message will be stored
  * @return  SUCCESS if a message is read
  */ 
-ErrorStatus MotorController_Read(CANbuff *message, car_state_t *car){
+ErrorStatus MotorController_Read(CANbuff *message){
 
     uint32_t id;
     uint8_t data[8] = {0};
-    uint32_t length = BSP_CAN_Read(CAN_3, &id, data);
     uint32_t firstSum = 0;
     uint32_t secondSum = 0;
+    CPU_TS ts;
+	OS_ERR err;
+	
+	// Check to see if a mailbox is available: BLOCKING
+	OSSemPend(&MotorController_ReceiveSem4,
+			  0,    
+			  OS_OPT_PEND_BLOCKING,
+			  &ts,
+			  &err);
+	assertOSError(0, err);
+    uint32_t length = BSP_CAN_Read(CAN_3, &id, data);
 
     if(length>0){
         message->id = id;
```

Comment:
Lines 133-154 could use some cleaning up, but it might not be the best use of our time.

For example, 
```C
        //get first number (bits 0-31)
        for(int j = 0; j < MAX_CAN_LEN/2; j++){
            firstSum += data[j];
            if(j != MAX_CAN_LEN/2 - 1){
                firstSum = firstSum << 8;
            }
        }
```
could be replaced by
```C
        //get first number (bits 0-31)
        for(int j = 0; j < MAX_CAN_LEN/2; j++){
            firstSum <<= 8;
            firstSum += data[j];
        }
```
and `union convert` doesn't look like it is used anywhere

Reply:
convert.f is used for velocity

Reply:
We could replace
```C
convert.n = secondSum;
CurrentVelocity = convert.f;
```
with
```C
memcpy(&CurrentVelocity, &secondSum, sizeof(CurrentVelocity));
```
or
```C
CurrentVelocity = *(float *) &secondSum;
```
and delete the union to make it more readable, but I'm fine with it if you leave it how it is.